src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1961,7 +1961,54 @@ let CheckMultipleInputsUsingGraphMode
 
         partialResults, tcState)
 
+let TryReuseTypecheckingResults (tcConfig: TcConfig) inputs =
+    let tcDataFileName = FileSystem.GetFullPathShim "FSharpTypecheckingData"
+    let nl = Environment.NewLine
+
+    let sourceFiles =
+        inputs
+        |> Seq.toArray
+        |> Array.mapi (fun idx (input: ParsedInput) ->
+            {
+                Idx = idx
+                FileName = input.FileName
+                ParsedInput = input
+            })
+
+    let filePairs = FilePairMap sourceFiles
+
+    let graph =
+        DependencyResolution.mkGraph filePairs sourceFiles
+        |> fst
+        |> Graph.map (fun idx ->
+            let friendlyFileName =
+                sourceFiles[idx]
+                    .FileName.Replace(tcConfig.implicitIncludeDir, "")
+                    .TrimStart([| '\\'; '/' |])
+
+            (idx, friendlyFileName))
+        |> Graph.asString
+
+    let thisTcData = $"{tcConfig.cmdLineArgs}{nl}{nl}{graph}"
+
+    if FileSystem.FileExistsShim tcDataFileName then
+        use tcDataFile = FileSystem.OpenFileForReadShim tcDataFileName
+        let existingTcData = tcDataFile.ReadAllText()
+
+        if thisTcData = existingTcData then
+            tcConfig.exiter.Exit 0
```

Comment:
In this phase, we don't cache any actual typechecking results yet... So when we detect there is no need for recompilation, there is nothing to restore and propagate yet.

So what do we do here? Just stop? That's weird.

Reply:
If the feature it cannot actually restore typechecking data, it should not do anything. Just signal a bool, ignore it, but do not exit.

Collecting metrics to actually evaluate the hit/miss ratios would be good as well.

Reply:
Well this we hit not when we cannot restore the data but we don't want to do that. Essentially, when we detect that nothing has changed since the last compilation. 

If we ignore it, we will keep recompiling code in such cases. Is that what we want at this stage?

Reply:
Mmm okay, removed the exit here. As for the metric - you mean collecting them via events?

Reply:
Some telemetry later telling us "how good" it is in terms of cache hits.

Reply:
That's clear, it's more about the mechanism to use. I only noticed events being used in the compiler which we could probably utilize. Or do you have some other idea?

Reply:
Use activities for now is good enough. Later on we'll hook up dotnet-collect compatible events

Reply:
Ah right, activities, that slipped from my head - thanks, will add.

Reply:
Added some activities there.